### PR TITLE
[TT-4147] remove APIID from bundle dest path

### PR DESCRIPTION
**Pull request by jeffy-mathew**
Currently bundle destpath is calculated with prefix `spec.APIID`, this could end up in different user experience for OSS users and dashboard users, since APIID is optional. 

This causes bundled plugins to be not used across more than 1 API. By removing the APIID prefix from bundle dest path, it'll not be dependant on APIID and hence more than 1 APIs can use the same plugin bundle

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3849/merge
Commit: 8eb5cd47e1421f5f87af327d3463c03a749880a8 
Triggered by: pull_request (@matiasinsaurralde)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1748238587)

**Comment by matiasinsaurralde** [comment]
@jeffy-mathew I've added some conditions in 8eb5cd47e1421f5f87af327d3463c03a749880a8, this plays well with existing bundles.

Regarding new bundles that haven't been downloaded yet, do we want to completely move to this new bundle naming approach? If so then the logic should be extended a bit:

1. Does the bundle already exist? For backwards compatibility we first check for `[API ID]_[Bundle hash]`, then we check for `[Bundle hash]`.
2.  If none of them exist then `getBundleDestPath` returns `[Bundle hash]` path.
2. If the bundle doesn't exist then the new approach is used and `[Bundle hash]` is used for all downloads.

From my tests 8eb5cd47e1421f5f87af327d3463c03a749880a8 plays well with two APIs that use the same Go plugin, given that you manually created a directory using `[Bundle hash]` as the middleware path but it doesn't take care of new bundles, so needs to be polished.

Now on the Python side, I'm looking at the middleware loader to see if re-using the directory in the context of a Python plugin would create any conflicts, Python module loading is pretty much based on paths and we add our own logic to be able to isolate different plugins and their internal dependencies.

**Comment by matiasinsaurralde** [comment]
@jeffy-mathew As supposed, the original change in cd10ef06d6da0820766f431b532b6def51a3073a breaks the Python module initialization code:
```
[Jan 26 03:09:47]  INFO main: ----> Fetching Bundle: python1.zip
[Jan 26 03:09:47]  INFO main: ----> Loading bundle: python1.zip
[Jan 26 03:09:47]  INFO main: ----> Verifying bundle: python1.zip
[Jan 26 03:09:47]  INFO main: ----> Bundle is valid, adding to spec: python1.zip
[Jan 26 03:09:47]  INFO coprocess: Python version '3.5' loaded
[Jan 26 03:09:48]  INFO python: Initializing dispatcher
[Jan 26 03:09:48]  INFO coprocess: Python dispatcher was initialized
[Jan 26 03:09:48]  INFO python: Loading module: '19bca2bdbcd398caf30b6803d289520e'
[Jan 26 03:09:48] ERROR python: PyObjectCallObject failed
[Jan 26 03:09:48] ERROR python: list index out of range
```
The reason is that the Python middleware loader uses a hook table -associated with API IDs- to dispatch the requests. The error is originated in [`load_bundle`](https://github.com/TykTechnologies/tyk/blob/master/coprocess/python/dispatcher.py#L37):

```python
    def load_bundle(self, bundle_path):
        path_splits = bundle_path.split('/')
        bundle_id = path_splits[-1]
        bundle = self.find_bundle(bundle_id)
        if not bundle:
            # TykMiddleware initialization fails here:
            bundle = TykMiddleware(bundle_id, bundle_root_path=self.bundle_root_path)
            self.bundles.append(bundle)
        self.update_hook_table(with_bundle=bundle)
```

If we dig deeper into [`TykMiddleware`](https://github.com/TykTechnologies/tyk/blob/master/coprocess/python/tyk/middleware.py#L14), we can find the cause of the `list index out of range` error:

```python
    def __init__(self, filepath, bundle_root_path=None):
        tyk.log( "Loading module: '{0}'".format(filepath), "info")
        self.filepath = filepath
        self.handlers = {}

        self.bundle_id = filepath
        self.bundle_root_path = bundle_root_path

        self.imported_modules = []
        
        # module_splits would only return a single element when no API ID is present on the bundle ID:
        module_splits = filepath.split('_')
        self.api_id, self.middleware_id = module_splits[0], module_splits[1]
```

So this is the reason why Python tests fail completely -good spot-, even after changes from 8eb5cd47e1421f5f87af327d3463c03a749880a8. I think that we should work on a workaround that ensures [`find_hook`](https://github.com/TykTechnologies/tyk/blob/master/coprocess/python/dispatcher.py#L62) is able to work with this new directory structure.

To summarize it's probably ok to rely only on bundle name hash as commit cd10ef06d6da0820766f431b532b6def51a3073a does. If a user was already using bundles, they will be freshly downloaded using the new directory structure. Note that different parts of the flow might be impacted by this change, first we have the bundle loading mechanism that occurs when a fresh gateway loads for the first time -let's say, without any existing bundles on disk-, secondly there's the Python middleware initialization process (established in `TykMiddleware` as shown above) and finally, when a request hits the gateway, we rely on `find_hook` to dispatch the request object to the proper hook.

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3849/merge
Commit: 15e4fd5c3e36f5b73c32c769ec95a1d07de91295 
Triggered by: pull_request (@matiasinsaurralde)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1757996215)

File: coprocess/python/tyk/middleware.py [hunk 1]
```diff
@@ -23,7 +23,11 @@ def __init__(self, filepath, bundle_root_path=None):
         self.imported_modules = []
         
         module_splits = filepath.split('_')
-        self.api_id, self.middleware_id = module_splits[0], module_splits[1]
+        self.api_id = ""
+        if len(module_splits) == 1:
+            self.middleware_id = module_splits[0]
+        else:
+            self.api_id, self.middleware_id = module_splits[0], module_splits[1]
```
Comment (jeffy-mathew) [inline coprocess/python/tyk/middleware.py:30]: do we need this block? we remove `api_id` prefix from bundle destpath [here](https://github.com/TykTechnologies/tyk/pull/3849/files#diff-72df0cbfd3765a5d0bff62196c11008596608a21a53dbb9c65bfc6f008dbfa29R322). so will there be cases where the filepath having `_` in it?

Comment (matiasinsaurralde) [reply coprocess/python/tyk/middleware.py:30]: Probably not required anymore, this is still WIP, need to run a few tests

Comment (matiasinsaurralde) [reply coprocess/python/tyk/middleware.py]: Removed this part

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3849/merge
Commit: 09f02dba11d55ca3095ed37ff593c9919c403191 
Triggered by: pull_request (@matiasinsaurralde)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1763515674)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3849/merge
Commit: ea84534ed4dc735843e1214bc80ed652f55a54ab 
Triggered by: pull_request (@matiasinsaurralde)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1763544640)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3849/merge
Commit: e0e43c58cee63940c10331b66328059f8a9a655b 
Triggered by: pull_request (@matiasinsaurralde)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1763592031)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3849/merge
Commit: c9a3e14cdc0b058066198e5f4c43b74bb3f6009e 
Triggered by: pull_request (@matiasinsaurralde)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1763625312)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3849/merge
Commit: 52220e82a9c10d3495979b205c9dd13060c7d3e3 
Triggered by: pull_request (@matiasinsaurralde)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1763720478)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3849/merge
Commit: bad595aabc1ecaaa6a06b18f828d1beee708e4b9 
Triggered by: pull_request (@matiasinsaurralde)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1772816206)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3849/merge
Commit: 63bdb40b2abbf3c52bcde535703fa9bfc9440358 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1772976729)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/3849/merge
Commit: 73c8542f1bc789573f05fe9daa616031d378ec5f 
Triggered by: pull_request (@jeffy-mathew)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1865481817)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/heads/master
Commit: 9124186d33217dcda14ede00ca5d60fd2a9dab91 [TT-4147] remove APIID from bundle dest path (#3849)

Currently bundle destpath is calculated with prefix `spec.APIID`, this could end up in different user experience for OSS users and dashboard users, since APIID is optional. 

This causes bundled plugins to be not used across more than 1 API. By removing the APIID prefix from bundle dest path, it'll not be dependant on APIID and hence more than 1 APIs can use the same plugin bundle
Triggered by: push (@buger)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/1865495155)
